### ISSUE TITLE
Streaming with server-sent events

### ISSUE THREAD

**Issue by gianlucatruda** [gianlucatruda/tai#1]

**Comment by gianlucatruda** [comment]
```bash
curl https://api.openai.com/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
    "model": "gpt-4o",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": "Hello!"
      }
    ],
    "stream": true
  }'
```

```
data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"role":"assistant","content":"","refusal":null},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":"Hello"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":"!"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":" How"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":" can"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":" I"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":" assist"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":" you"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":" today"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{"content":"?"},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc","object":"chat.completion.chunk","created":1727196805,"model":"gpt-4o-2024-05-13","system_fingerprint":"fp_52a7f40b0b","choices":[{"index":0,"delta":{},"logprobs":null,"finish_reason":"stop"}]}

data: [DONE]

```

```json
{
  "id": "chatcmpl-AB34vbMH9lDlSR3VfLZABnDm6DnCc",
  "object": "chat.completion.chunk",
  "created": 1727196805,
  "model": "gpt-4o-2024-05-13",
  "system_fingerprint": "fp_52a7f40b0b",
  "choices": [
    {
      "index": 0,
      "delta": {
        "role": "assistant",
        "content": "",
        "refusal": null
      },
      "logprobs": null,
      "finish_reason": null
    }
  ]
}
```

**Comment by gianlucatruda** [comment]
```json
{
  "id": "chatcmpl-AB7avQOtBvfeu3WPdboAKVtofdZXQ",
  "object": "chat.completion.chunk",
  "created": 1727214165,
  "model": "gpt-4o-2024-05-13",
  "system_fingerprint": "fp_e375328146",
  "choices": [
    {
      "index": 0,
      "delta": {
        "role": "assistant",
        "content": "",
        "refusal": null
      },
      "logprobs": null,
      "finish_reason": null
    }
  ]
}

```
```json
{
  "id": "chatcmpl-AB7avQOtBvfeu3WPdboAKVtofdZXQ",
  "object": "chat.completion.chunk",
  "created": 1727214165,
  "model": "gpt-4o-2024-05-13",
  "system_fingerprint": "fp_e375328146",
  "choices": [
    {
      "index": 0,
      "delta": {
        "content": " How"
      },
      "logprobs": null,
      "finish_reason": null
    }
  ]
}

```
```json
{
  "id": "chatcmpl-AB7avQOtBvfeu3WPdboAKVtofdZXQ",
  "object": "chat.completion.chunk",
  "created": 1727214165,
  "model": "gpt-4o-2024-05-13",
  "system_fingerprint": "fp_e375328146",
  "choices": [
    {
      "index": 0,
      "delta": {},
      "logprobs": null,
      "finish_reason": "stop"
    }
  ]
}
```

**Comment by gianlucatruda** [comment]
Fixed in [88fb2a4a097fd5abcc3900229e983c80f9ef1db8](https://github.com/gianlucatruda/tai/commit/88fb2a4a097fd5abcc3900229e983c80f9ef1db8)
